### ISSUE TITLE
The code of Short-range Consistency and Long-range Consistency

### ISSUE THREAD

**Issue by HeChengy** [xingy038/ConRF#1]
Hello, thank you for your excellent research. Can you share the code for run-related metrics? Can you share the code of Short-range Consistency and Long-range Consistency? thank you.

**Comment by xingy038** [comment]
import numpy as np
    import torch
    import lpips
    from PIL import Image
    import scipy

    def rgb_ssim(img0, img1, max_val, filter_size=11, filter_sigma=1.5, k1=0.01, k2=0.03, return_map=False):
        hw = filter_size // 2
        shift = (2 * hw - filter_size + 1) / 2
        f_i = ((np.arange(filter_size) - hw + shift) / filter_sigma)**2
        filt = np.exp(-0.5 * f_i)
        filt /= np.sum(filt)
    
        def convolve2d(z, f):
            return scipy.signal.convolve2d(z, f, mode='valid')
    
        filt_fn = lambda z: np.stack([
            convolve2d(convolve2d(z[...,i], filt[:, None]), filt[None, :])
            for i in range(z.shape[-1])], -1)
        mu0 = filt_fn(img0)
        mu1 = filt_fn(img1)
        mu00 = mu0 * mu0
        mu11 = mu1 * mu1
        mu01 = mu0 * mu1
        sigma00 = filt_fn(img0**2) - mu00
        sigma11 = filt_fn(img1**2) - mu11
        sigma01 = filt_fn(img0 * img1) - mu01
    
        sigma00 = np.maximum(0., sigma00)
        sigma11 = np.maximum(0., sigma11)
        sigma01 = np.sign(sigma01) * np.minimum(
            np.sqrt(sigma00 * sigma11), np.abs(sigma01))
        c1 = (k1 * max_val)**2
        c2 = (k2 * max_val)**2
        numer = (2 * mu01 + c1) * (2 * sigma01 + c2)
        denom = (mu00 + mu11 + c1) * (sigma00 + sigma11 + c2)
        ssim_map = numer / denom
        ssim = np.mean(ssim_map)
        return ssim_map if return_map else ssim

    image1_path = ''
    image1 = Image.open(image1_path)
    image1 = image1.resize([W, H])
    
    image2_path = ''
    image2 = Image.open(image2_path)
    image2 = image2.resize([W, H])
    
    image1_np = np.array(image1)
    image2_np = np.array(image2)
    
    loss_fn = lpips.LPIPS(net='alex')  
    
    image1_tensor = torch.Tensor(image1_np)[None, ...].permute(0, 3, 1, 2)
    image2_tensor = torch.Tensor(image2_np)[None, ...].permute(0, 3, 1, 2)
    
    lpips_score = loss_fn(image1_tensor, image2_tensor)
    ssim = rgb_ssim(image1_tensor.permute(0,2,3,1)[0], image2_tensor.permute(0,2,3,1)[0], 1)
    
    print(f"ssim: {ssim}")
    print(f"LPIPS Score: {lpips_score.item()}")

**Comment by xingy038** [comment]
To get warping t->t+1, you need to get optical flow first (using the pre-trained model, such as RAFT), and then use the following code.

    def warp(x, flo):
        """
        warp an image/tensor (im2) back to im1, according to the optical flow
        x: [B, C, H, W] (im2)
        flo: [B, 2, H, W] flow
        """
        B, C, H, W = x.size()
        # mesh grid
        xx = torch.arange(0, W).view(1, -1).repeat(H, 1)
        yy = torch.arange(0, H).view(-1, 1).repeat(1, W)
        xx = xx.view(1, 1, H, W).repeat(B, 1, 1, 1)
        yy = yy.view(1, 1, H, W).repeat(B, 1, 1, 1)
        grid = torch.cat((xx, yy), 1).float()
    
        if x.is_cuda:
            grid = grid.cuda()
        vgrid = grid + flo
        # scale grid to [-1,1]
        vgrid[:, 0, :, :] = 2.0 * vgrid[:, 0, :, :].clone() / max(W - 1, 1) - 1.0
        vgrid[:, 1, :, :] = 2.0 * vgrid[:, 1, :, :].clone() / max(H - 1, 1) - 1.0
    
        vgrid = vgrid.permute(0, 2, 3, 1)
        output = F.grid_sample(x, vgrid)
        mask = torch.ones(x.size()).to('cuda')
        mask = F.grid_sample(mask, vgrid)
    
        mask[mask < 0.999] = 0
        mask[mask > 0] = 1
    
        return output * mask

**Comment by HeChengy** [comment]
Thank you. And, how far exactly is "long-range consistency"? How should it be set?

**Comment by xingy038** [comment]
> Thank you. And, how far exactly is "long-range consistency"? How should it be set?

7 frame intervals.

Please refer to this paper "https://people.cs.nycu.edu.tw/~walon/publications/chiang2022wacv.pdf".
